### ISSUE TITLE
[bug] logprobs & entropy mismatch between w/ and w/o enabling kernel fusion

### ISSUE THREAD

**Issue by WindowsXp-Beta** [verl-project/verl#2656]
We have been noticing reward collapse after enabling linear kernel fusion. Through some logging we found that it was caused by the mismatch of `logprobs` and `entropy`(though it's just used in metrices) between w/ and w/o enabling kernel fusion. Specifically, taking one of our training  jobs as an example, the `logprobs` calculated by `kernels.efficient_entropy_forward` in `utils/kernel/linear_cross_entropy.py` is:
https://github.com/volcengine/verl/blob/ac826e0558017a9c675818f36cbb3473c14a2a50/verl/utils/kernel/linear_cross_entropy.py#L77-L79
```
> logprobs.max(), logprobs.min(), logprobs.mean()
(tensor(123.7357, device='cuda:0'), tensor(-27.1481, device='cuda:0'), tensor(59.2606, device='cuda:0'))
```
which doesn't make sense since logprobs should be less than 0.
We also calculated the `logprobs` following `run_verl_original_entropy` in `tests/utils/test_linear_cross_entropy.py` and the result was
https://github.com/volcengine/verl/blob/ac826e0558017a9c675818f36cbb3473c14a2a50/tests/utils/test_linear_cross_entropy.py#L64
```
> logprobs_verl.max(), logprobs_verl.min(), logprobs_verl.mean()
(tensor(-0., device='cuda:0'), tensor(-27.1313, device='cuda:0'), tensor(-1.2149, device='cuda:0'))
```
The entropy differed as well(the only difference between the two traces below is whether or not using fused kernel).
<img width="457" height="327" alt="Image" src="https://github.com/user-attachments/assets/3b1e47b4-e127-4248-9438-f254fb62cad8" />

cc @Jianbing-D @ETOgaosion

**Comment by WindowsXp-Beta** [comment]
Another interesting observation, I manually called `linear_cross_entropy` multiple times in a debug session and the results were different🤔:

<img width="854" height="368" alt="Image" src="https://github.com/user-attachments/assets/1da75d97-9b80-4d26-b2f1-72af6611c346" />

**Comment by WindowsXp-Beta** [comment]
After more debugging, we have found that:
1. The indeterminism of `linear_cross_entropy` and having positive value in calculated logprobs were caused by our model having negative token id. After we `masked_fill` out those tokens, the results became deterministic and fell into the right range.
2. However, the results still mismatched between `linear_cross_entropy` and vanilla pytorch implementation. We have narrowed down the problem to be that the triton kernel is not stable on our `hidden_states` and `weight` values. Specifically, we `torch.load` the hidden_states and weights into https://github.com/volcengine/verl/blob/ac826e0558017a9c675818f36cbb3473c14a2a50/tests/utils/test_linear_cross_entropy.py#L182
And `kernel_logprobs` and `verl_logprobs` differed significantly:
    ```
    verl_kernel_diff = torch.nonzero(~torch.isclose(verl_logprobs, kernel_logprobs, atol=1e-1))
    verl_kernel_diff.shape
    > torch.Size([7923, 1])
    ```
    We have uploaded two tensors to [Google Drive](https://drive.google.com/drive/folders/1KHD7AYKeNPMomruXj4AKAqt_DdkqHGPM?usp=sharing). Anyone could reproduce it easily.

**Comment by vadimkantorov** [comment]
cc @vermouth1992 @Jianbing-D

**Comment by vadimkantorov** [comment]
@WindowsXp-Beta could you please publish a complete repro script demonstrating the problem? (maybe loading the inputs from the pt files you provided)

**Comment by WindowsXp-Beta** [comment]
@vadimkantorov sure, this a single python file that extracts minimal functions from `test_linear_cross_entropy.py` to reproduce the mismatch. [reproduce_triton_kernel_mismatch.py.txt](https://github.com/user-attachments/files/21483421/reproduce_triton_kernel_mismatch.py.txt)

I used `gdown` to download files from Google Drive.
```
pip install gdown
gdown https://drive.google.com/uc\?id\=1YGXu2ZpQTZYthXmbKSxH3JdwalMjvnfh # download hidden_states.pt
gdown https://drive.google.com/uc\?id\=10mBCx_ixNriZ6S4DNeYUMypXINCR7Mzg # download weights.pt
```

**Comment by vadimkantorov** [comment]
I ran your repro. And also confirm that if I add `hidden = torch.rand_like(hidden)` or `weight = torch.rand_like(weight)`, the mismatch disappears.

So on my end I also confirm that it's interaction of your `hidden`, `weight` which triggers the problem. Maybe some sort of overflow in Triton?

I'm trying to debug

**Comment by vadimkantorov** [comment]
@Jianbing-D could this be an issue with some intermediate dtype casting?

**Comment by vadimkantorov** [comment]
@Jianbing-D also, if `entropy_coef == 0 and calculate_entropy == False`, could directly Liger triton kernel be used?

**Comment by vadimkantorov** [comment]
@WindowsXp-Beta I've studied the repro. I think:

1. The fused triton implementation in Verl accepts the `hidden`, `vocab` in original `bfloat16` format, does the tile matmul in bfloat16 (that is the multiplications are done in bfloat16), and only does accumulation into `float32` logits. For apples-to-apples comparison, `hidden` and `weight` tensors must be converted to `float32` before calling `linear_cross_entropy`
2. The reference function `logprobs_from_logits`'s code path depends on presence of FlashAttentio, so maybe better for the purpose of this check to set `verl.utils.torch_functional.FLAH_ATTN_CROSS_ENTROPY_LOSS_AVAILABLE = False` to simplify the repro

2. I've tested `run_verl_torch_fused_entropy`, it works fine

3. Even if I replaced the `linear_cross_entropy(hidden, weight, labels, temperature)` with `linear_cross_entropy(hidden.to(torch.float32), weight.to(torch.float32), labels, temperature)` and set `labels` to `labels = torch.zeros(batch_size, num_tokens, dtype = torch.int64, device = 'cuda')`, I still get the discrepancy and high max-abs-diff (of 10.3296), so there is indeed some problem with the fused Triton kernel on these inputs (which goes to almost zero if `hidden` or `weight` are replaced with `torch.rand_like(hidden)`), cc @Jianbing-D @vermouth1992 

4. If entropy `entropy_coef == 0 and calculate_entropy == False, could directly use the fused_linear_cross_entropy Liger kernel, right?`
---

I'll share more if I find the problem in the Triton kernel. If not related to upcasting - I wonder, what could be the reason.

**Comment by WindowsXp-Beta** [comment]
Re: 5 => Doesn't the linger kernel only support calculating entropy? https://github.com/volcengine/verl/pull/462#issuecomment-2786132409

**Comment by vadimkantorov** [comment]
> Re: 5 => Doesn't the linger kernel only support calculating entropy?

@WindowsXp-Beta I think liger kernel only supports calculating cross-entropy (consuming labels). But when `entropy_coef == 0 and calculate_entropy == False`, it should be sufficient

**Comment by vadimkantorov** [comment]
(I'm currently suspecting the bug/difference from reference might be in how the stabilized logsumexp is calculated / online local-global maximum's are calculated)

**Comment by vadimkantorov** [comment]
Confirmed, the problem unsurprisingly is in calculated logsumexp values (which are subtracted from logits)

(cc @Jianbing-D @vermouth1992)

**Comment by WindowsXp-Beta** [comment]
> Confirmed, the problem unsurprisingly is in calculated logsumexp values (which are subtracted from logits)
> 
> (cc [@Jianbing-D](https://github.com/Jianbing-D) [@vermouth1992](https://github.com/vermouth1992))

I have shared your finding to the verl community channel in [lark](https://www.larksuite.com/en_sg/) which should have better visibility.

**Comment by vadimkantorov** [comment]
If I understand well, verl uses two-level logsumexp calculation - is it for increasing parallelism? (as opposed to Liger / Triton which do single-level)

Also, regular PyTorch in eager uses two-pass impl (which first computes maxima, and then does another memory pass and computes the aggregation):
- https://github.com/pytorch/pytorch/issues/31837

**Comment by shimizust** [comment]
What version of liger are you using?

**Comment by vadimkantorov** [comment]
Could it be that the precision is deteriorated because of two-level aggregation? (and thus multiplication of exp's)?

Because otherwise, it just seems that the algorithm is just two-level application of https://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html

---

Could verl also support the slower, but more reliable single-level aggregation implementation like Liger or FlashAttn?

https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/cross_entropy.py

https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/ops/triton/cross_entropy.py

**Comment by vadimkantorov** [comment]
@WindowsXp-Beta, it also appears that your repro input can be just reduced to one-two tokens:
```python
hidden = hidden[:, 6202:6202+2, :].to(torch.float32).contiguous()
labels = labels[:, 6202:6202+2].contiguous()
```

Also, the issue is not with Triton per se, as the same discrepancy can be observed when run under `TRITON_INTERPRET=1` which interprets/executes the Triton program in sequential way on CPU using NumPy.

I think the problem lies with the precision of the two-level aggregation algorithm on some inputs.

**Comment by WindowsXp-Beta** [comment]
@vadimkantorov wow, that's a good catch. I'll check more hidden states.

**Comment by vadimkantorov** [comment]
@WindowsXp-Beta I think I found the culprit, the Triton impl assumes vocab divisibility by 256 and 1024:
- https://github.com/volcengine/verl/issues/2899

And in your example it's only divisible by 128, so the padded dimensions poison the logsumexp computation, when the hidden scale is larger than the something close to zero.

**Comment by vadimkantorov** [comment]
@WindowsXp-Beta could you please confirm if indeed replacing 256/1024 with 128 in kernels.py (linked in https://github.com/volcengine/verl/issues/2899) fixes the problem?

**Comment by WindowsXp-Beta** [comment]
Hi @vadimkantorov , sorry for the late response. I was occupied by other things these days. Thanks so much for digging into the problem and making such amazing findings! I tested your suggested solution on the repro script and it worked! I have launched a training job and will compare the reward/loss with the run without using fused kernel.
But what I don't understand is that if I use the same vocab_size(129920) to rand `hidden_states`(uniform_(-25, 25)) and `weight` tensors, why they don't lead to discrepancy?

**Comment by vadimkantorov** [comment]
> I have launched a training job and will compare the reward/loss with the run without using fused kernel.

Thank you for uploading the bad input in the first place!

> why they don't lead to discrepancy?

My understanding is that the scale of `hidden` matters. Maximum value for your tensor is about 50. If it's smaller (for random `hidden`), then the average value of "true" logits is smaller, and diverges less from padded zero logits.

But I might not fully understand what's going on. E.g. my example of doing only logsumexp https://gist.github.com/vadimkantorov/19a6f22930705ed9a6497192eb44f304 on logits (without gemm), works fine even with 1024/256. So if you'd like to dig further, one would need to understand what's the difference of the code (wrt padding) just doing the logsumexp and of the code doing gemm+logsumexp...

I guess we'd need some input from @Jianbing-D @vermouth1992 who wrote down this Triton kernel

**Comment by wuxibin89** [comment]
There may be potential bug in trition kernel implementaion: https://github.com/volcengine/verl/pull/462#issuecomment-3713968824
